src/lib.rs

```diff
@@ -346,10 +349,326 @@ impl From<Error> for io::Error {
     }
 }
 
+/// Trait for path-like types.
+///
+/// This trait defines most of the operations one might want to perform on a
+/// path-like type. It very closely resembles the interface of the standard
+/// library's [`PathBuf`], except that methods that can return an error will
+/// return a rich [`path_abs::Error`] rather than a plain [`std::io::Error`].
+///
+/// Trait methods without documentation should be assumed to be identical to the
+/// method of the same name on [`PathBuf`].
+///
+/// [`PathBuf`]: https://doc.rust-lang.org/stable/std/path/struct.PathBuf.html
+/// [`path_abs::Error`]: struct.Error.html
+/// [`std::io::Error`]: https://doc.rust-lang.org/stable/std/io/struct.Error.html
+pub trait PathExt: Sized + Clone + Borrow<PathBuf> + From<PathBuf> + Into<PathArc> {
```

Comment:
I think we should not require `From<PathBuf>`, otherwise we aren't going to be able to implement for `PathFile`, etc.

Reply:
`From<PathBuf>` is so that we can turn the result of (say) `Path::canonicalize()` back into whatever type the trait is implemented for. If you called `parent()` on a `PathAbs`, I think you'd expect to get back a new `PathAbs`, not a `PathBuf`.

Alternatively, we could say that `PathFile` is *just* for file-manipulation functions, and shouldn't support the arbitrary path-manipulation methods of `PathExt` at all. Which seems reasonable to me; if you call `.concat()` or `.with_file_name()` on a `PathFile`, what you get back might not be a valid, existing `PathFile` after all.

Perhaps the trait could be split into "methods that return information about the current path" (implemented everywhere), and "methods that return new, related paths" (implemented for types that are allowed to hold generic paths)?

Reply:
An update: I've been trying to replace `PathArc` with`PathExt`, but not implementing `PathExt` for `PathFile` and `PathDir` to avoid breaking their invariants. It's gone pretty well, except that there's a bunch of things that want to call `.display()` on them, or call `.join()` and get back a `PathAbs`.

I'm thinking I should break `PathExt` into `PathInfo` (`.is_absolute()`, `.display()`, etc.) and `PathOps` (`.concat()`, `.canonicalize()`, etc.) traits, where `PathInfo` is implemented for maybe `T: AsRef<Path>` and `PathOps` is implemented for `PathBuf`, `Arc<PathBuf>` and `PathAbs`. What do you think about that?